### ISSUE TITLE
[$250] Chat - After editing nested quote message, the markdown is removed

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#47951]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.24-1
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/4884121
**Email or phone of affected tester (no customers):** applausetester+vd_web082324@applause.expensifail.com
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Go to https://staging.new.expensify.com/
2. Log in
3. Go to any chat
4. Send a nested quote message, like "> > test"
5. Verify that it looks as expected on the chat
6. Edit the message

## Expected Result:
The markdown should still be displayed after editing the message

## Actual  Result:
The quote markdown effect is removed after editing a nested quote message

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/user-attachments/assets/3a9ad135-4bdb-4575-9683-e8f2a9a03643

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01a7ea8bf9646df697</li>
        <li>Upwork Job ID: 1829295392648638850</li>
        <li>Last Price Increase: 2024-08-29</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @fedirjh</details>

**Comment by lanitochka17** [comment]
@twisterdotcom  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by lanitochka17** [comment]
We think that this bug might be related to #vip-vsp

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
After editing a nested quote markdown, the quote markdown is removed.

### What is the root cause of that problem?
When we send `> > test` and edit the message, the markdown becomes `>> test` because we add the `>` without space and only add the space after the `>`. We can change that but I want to focus on a real issue explained below.
https://github.com/Expensify/expensify-common/blob/d11466167bc562447d2dfdad624145040153efae/lib/ExpensiMark.ts#L624

A quote markdown requires a space after `>`, for example, `> test`. But if we have a nested quote, `>> test`, there is an inconsistent logic when parsing it. If we see the quote markdown regex, we allow the nested quote without spaces between `>`, so `>> test` is a valid nested quote.
https://github.com/Expensify/expensify-common/blob/d11466167bc562447d2dfdad624145040153efae/lib/ExpensiMark.ts#L376

`modifyTextForQuote` is responsible to process the matched text (`>> test`) into the HTML form. It splits the text by new line and process each line whether it should be converted to a quote HTML or not. On each line, it checks whether the line starts with `> ` or not.
https://github.com/Expensify/expensify-common/blob/d11466167bc562447d2dfdad624145040153efae/lib/ExpensiMark.ts#L1138

This shows that the logic doesn't cover the case of nested quote yet. So, you don't need to edit the message but just sends `>> test` and it won't be converted to a quote (the live markdown converts it to quote though because it doesn't get through `modifyTextForQuote` processing).

### What changes do you think we should make in order to solve the problem?

To allow nested quote have optional space between `>`, we can update the logic to accept either space or another `>` after `>` (we can use regex to shorten it, `^&gt;(&gt;| )`)
```
if ((Str.startsWith(textSplitted[i], '&gt; ') || Str.startsWith(textSplitted[i], '&gt;&gt;') || isLastBlockquote) && !insideCodefence) {
```
This will convert it to markdown for below inputs:
```
> > test
>> test
```

**Comment by ChavdaSachin** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Editing quote markdown message does not work as desired

### What is the root cause of that problem?
RCA for the issue is -
For comments with markdown, we store HTML equivalent of markdown comments in database.
So, when we try to edit a comment - first we convert the comment from HTML equivalent to markdown equivalent comment for editing using [htmlToMarkdown](https://github.com/Expensify/expensify-common/blob/d11466167bc562447d2dfdad624145040153efae/lib/ExpensiMark.ts#L1073-L1081) function provided by [expensify-common](https://github.com/Expensify/expensify-common) project of ours. When converting comment with `quote`  back to markdown, we use [this rule](https://github.com/Expensify/expensify-common/blob/d11466167bc562447d2dfdad624145040153efae/lib/ExpensiMark.ts#L599-L631) which has a flaw and does not provide us with exact equivalent markdown.

### What changes do you think we should make in order to solve the problem?
Changing  [this rule](https://github.com/Expensify/expensify-common/blob/d11466167bc562447d2dfdad624145040153efae/lib/ExpensiMark.ts#L599-L631)  slightly would provide correct markdown equvalent.
Change [Here](https://github.com/Expensify/expensify-common/blob/d11466167bc562447d2dfdad624145040153efae/lib/ExpensiMark.ts#L624) from -
```javascript
return `${'>'.repeat(depth)} ${modifiedText}`;
```
To -
```javascript
return `${'> '.repeat(depth)}${modifiedText}`;
```
changed blankspace position.
<details>
  <summary> <h2>Result</h2></summary>
  
https://github.com/user-attachments/assets/648c8114-cf79-4a5a-9b3a-4b34c52ab485

  </details>

### What alternative solutions did you explore? (Optional)

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by twisterdotcom** [comment]
Agree, this one was simple to reproduce. Apologies for the time, I am just back from OOO.

**Comment by ChavdaSachin** [comment]
[updated proposal](https://github.com/Expensify/App/issues/47951#issuecomment-2308868522), added alternative solution which looks more sensible to me

**Comment by twisterdotcom** [comment]
Bump @fedirjh on these proposals

**Comment by ChavdaSachin** [comment]
@fedirjh It looks like two blockquotes must be separated by space is recommended design feature. So, we might need a design review on this one. check https://github.com/Expensify/App/issues/45154#issuecomment-2222162593 .

**Comment by bernhardoj** [comment]
> When we send > > test and edit the message, the markdown becomes >> test because we add the > without space and only add the space after the >. We can change that but I want to focus on a real issue explained below.
https://github.com/Expensify/expensify-common/blob/d11466167bc562447d2dfdad624145040153efae/lib/ExpensiMark.ts#L624

Just a note that if we require space for each `>`, I already write an explanation about that in the root cause section.

**Comment by ChavdaSachin** [comment]
## Proposal
[updated](https://github.com/Expensify/App/issues/47951#issuecomment-2308868522)
@fedirjh would you please take a look at my updated proposal. since we are missing out on multiline quote problem without completely modifying HTMLToMarkdown rule for 'quote' as I suggested

https://github.com/user-attachments/assets/32296991-dc32-484d-adee-166947d9e1e0

**Comment by twisterdotcom** [comment]
Bump on this @fedirjh.

**Comment by cead22** [comment]
Proposals are still being reviewed

**Comment by cead22** [comment]
Proposals are still being reviewed

**Comment by cead22** [comment]
Same same, and I'm going OOO tomorrow, in case you need to find another engineer @twisterdotcom

**Comment by twisterdotcom** [comment]
cc @BrtqKr Friendly bump for https://github.com/Expensify/App/issues/47951#issuecomment-2339349412

**Comment by twisterdotcom** [comment]
Bumped in Slack: https://expensify.slack.com/archives/C06BDSWLDPB/p1727971180783039

**Comment by twisterdotcom** [comment]
Bumped slack

**Comment by ChavdaSachin** [comment]
@fedirjh please make sure to think about the second problem I have mentioned in my proposal
> - when editing multiline quotes, when preceding line have more quote than following line, no '>' for following line is returned
It may be stale but why not to eliminate the if we get chance, and my proposal exactly does that.

**Comment by twisterdotcom** [comment]
I agree, I would love to solve this completely. @BrtqKr and @fedirjh what do you think of the proposal by @ChavdaSachin?

@ChavdaSachin could you confirm you've looked at the above issues from @BrtqKr too and incorporated anything relevant into your proposed fix?

**Comment by ChavdaSachin** [comment]
@twisterdotcom we have few updates in expensi common file after my proposals, I would look into it and make a changes needed in short time. and at the time I wrote my proposal I had all these case on my mind and they were working fine.

**Comment by ChavdaSachin** [comment]
[ExpensiMark.js](https://github.com/ChavdaSachin/ModifiedFiles/blob/ebaee8cd6a60405c7bfe5501df8f82f7ce2e3cbd/fix%203/ExpensiMark.js) this is modified version of ExpensiMark.js file. (file is of old version, I will implement the same on new version once I get green flag on behavior)
All quote rules match with current staging env with fix for the current issue.

Please checkout on dev env using this file.
And let me know if we want any other improvements.

cc. @twisterdotcom @BrtqKr @fedirjh

**Comment by Skalakid** [comment]
@ChavdaSachin Changes make sense to me, and everything works fine on the live markdown side. However, on the Expensify App side, the multiline quotes don't work. It might be connected to the changes that were made after your proposal was posted. It would be nice to make the message styling consistent with the styles in live markdown input  

https://github.com/user-attachments/assets/bcbd3a84-c7a2-447d-9fb1-2c5da5ff0d5d

**Comment by cead22** [comment]
Proposals are still being evaluated ^

**Comment by cead22** [comment]
Same same ^

**Comment by twisterdotcom** [comment]
@ChavdaSachin - would love your thoughts @Skalakid's reply please.

**Comment by cead22** [comment]
@ChavdaSachin bump ^

**Comment by ChavdaSachin** [comment]
Hey @Skalakid, it seems like you are missing out on something here.
- It's been a while, we have decided to consider only `>`s without space in-between as invalid multi quote.
- see this [Reference Comment](https://github.com/Expensify/App/issues/47951#issuecomment-2396835453)

But if we want to move ahead and match the behavior with live Markdown(ie. multiple `>`s without space in-between) I can for sure do that as well, in fact it was part of my initial proposal.

On the other hand I am also down if we want to make changes to live-MD and only allow multiple `>`s with space in-between as a valid case.

thoughts?

cc. @Skalakid @cead22 @twisterdotcom @BrtqKr

**Comment by Skalakid** [comment]
Oh okay, so in this case I think we should focus only on allowing spaces between `>`s  while editing a message, like is was said https://github.com/Expensify/App/issues/47951#issuecomment-2396835453

**Comment by Skalakid** [comment]
@ChavdaSachin, can you please update your proposal so it can be applied to the current code of E/App? I can't make it work. I will verify if everything works fine, and I think we will continue with your proposal

**Comment by ChavdaSachin** [comment]
Sounds good,

here's my plan - 
- first I will prepare a document on all the expected behavior, and post the document here for approval.
as soon as you guys approve the document.
- I will come up with updated rule which follows the document.

Agree ?

**Comment by twisterdotcom** [comment]
How is this going @ChavdaSachin?

**Comment by ChavdaSachin** [comment]
I was a little occupied these days, would complete the doc within this weekend.

**Comment by cead22** [comment]
Waiting for the doc ^ but it sounds like it'll be ready soon

**Comment by ChavdaSachin** [comment]
These tests cover most of the cases with quote MD.
Please take a look.
I particularly think we need to discuss tests: 1-5, 12-13, 21, 26-27.
[Quote MD Tests.docx](https://github.com/user-attachments/files/17546543/Quote.MD.Tests.docx)

Test text - 
[Quote tests 1-27.odt](https://github.com/user-attachments/files/17546698/Quote.tests.1-27.odt)

GDrive link for the document.(We could probably add comments to the document for discussion.)
https://docs.google.com/document/d/1by99UuFQ38QnApljTvlMWUtveB4EpZ-F/edit?usp=sharing&ouid=103248957501772037676&rtpof=true&sd=true

cc. @cead22 @twisterdotcom @Skalakid @BrtqKr @fedirjh

**Comment by cead22** [comment]
@ChavdaSachin @fedirjh can you start a discussion in #expensify-open-source about the desired behavior, and maybe share here what you would propose it should be? Thanks

**Comment by cead22** [comment]
@ChavdaSachin @fedirjh can you start a discussion in #expensify-open-source about the desired behavior, and maybe share here what you would propose it should be? Thanks

**Comment by ChavdaSachin** [comment]
I am not yet invited to slack, so that would not be possible for me currently...

**Comment by cead22** [comment]
@ChavdaSachin Can you break down the options here for me, and I can start the slack discussion?

**Comment by cead22** [comment]
@ChavdaSachin Can you break down the options here for me, and I can start the slack discussion?

**Comment by ChavdaSachin** [comment]
# let me quickly summarize this issue here...
- Our goal for this issue is to solve the problems which arise when editing existing markdown comment containing quote markdown.

- In order to archive that, We have to refactor html to mark-down rule and completely rewrite the markdown to html rule for quote markdown**.

- Since we are redesigning the quote rule, we have to take care that any existing functionalities/test should not fail and also we do not carry forward any existing bug.

- So first sub goal here is to go through the doc I shared [here](https://github.com/Expensify/App/issues/47951#issuecomment-2442218827) and decide each test is producing expected result or contains bug.
- As soon as we pick all the bug from current rule, we could proceed on writing new rule.

<details><summary> ** Reason behind complete rewrite:</summary>

For multiquote markdown to html, html quotes are not assigned line by line.
ie. 
```
> > a
> b
```
Result - 
```
<quote><quote>a
</quote>b</quote>
```
Here single line does not contain all the require context about depth.
So new rule should produce results line by line so no need to have any context about depth and we could directly apply greedy algorithm line by line.

Result to be produced by the new rule:
```
<quote><quote>a</quote></quote>
<quote>b</quote>
```

</details>

**Comment by twisterdotcom** [comment]
Hi @ChavdaSachin - we'll assign Michal here but thanks so much for your effort. If you could email contributors@expensify.com again, we should be able to get you into Slack for future issues.

**Comment by ChavdaSachin** [comment]
@twisterdotcom I will mail again, but assigning someone else here sounds a bit unfair after so much efforts I had here....
all I need right now is just clear expectations

**Comment by cead22** [comment]
Thanks @bernhardoj @ChavdaSachin @fedirjh @BrtqKr @Skalakid for your help here. @Skalakid is going to take over this issue as SWM has experience with live markdown and they're working on other similar bugs ([eg](https://github.com/Expensify/App/issues/45154))

**Comment by Skalakid** [comment]
### Update

Hi, today I analyzed the code, and I found that the bug is caused by the different `blockquote` parsing logic for the app-sent messages and the Live Markdown Input. For sending messages, we need to remove `>` characters, and if one is under another, we need to group them into one bigger `blockquote`. Thanks to that, our app can easily render these quotes as one without any space between them. Something like this:

![Screenshot 2024-11-14 at 16 57 13](https://github.com/user-attachments/assets/cf5819cc-7c82-4317-a680-3272abe5f1ab)

In the Live Markdown, we don't need to group these tags since the styles are a bit different, and there is a different logic behind rendering quotes.

So, to solve the problem, we need to unify the way how `blockquotes` are parsed for both destinations. Thanks to that, we won't have any problems with differences between Live Markdowns Input and markdown rendering. We can do it by refactoring the `replacement` and `rawInputReplacement` functions and replacing the `modifyTextForQuote` and `formatTextForQuote` functions with a different logic.
I will get back to it tomorrow

**Comment by cead22** [comment]
Work is happening and we have a draft PR

**Comment by Skalakid** [comment]
Hello, Here is the https://github.com/Expensify/expensify-common/pull/820 with the fix. I changed the blockquote parsing logic and now I'm working on changing html to markdown conversion algorithm

**Comment by Skalakid** [comment]
Hello, today I continued fixing HTML to markdown parsing logic. Tomorrow I will clean up the code and push the changes to the PR so we can start merging it

**Comment by Skalakid** [comment]
Hello, I'm adding last touches to my [PR](https://github.com/Expensify/expensify-common/pull/820) with the solution for this issue. Will open it for the review today

**Comment by Skalakid** [comment]
[PR](https://github.com/Expensify/expensify-common/pull/820) is waiting for the C+ review

**Comment by twisterdotcom** [comment]
This was merged, so it is waiting on a deploy now?

**Comment by twisterdotcom** [comment]
Would be nice to get an update on this one. Is it likely be getting daily updates, or should we make it weekly for now if we're beholden to some other deployment schedules?

**Comment by Skalakid** [comment]
The solution is ready; we are only waiting to bump the libraries in the app. As I wrote above today, this will be done by this [PR](https://github.com/Expensify/App/pull/53627). For now, we can make it weekly, but the changes should be merged this week if nothing critical appear

**Comment by twisterdotcom** [comment]
Wait, with no payment?
